keps/sig-api-machinery/20200421-static-admission-webhook-configuration.md

```diff
@@ -0,0 +1,149 @@
+policy addition and kube-apiserver startup. 
+
+## Motivation
+
+Today most policy enforcement is implemented through MutatingAdmissionWebhooks
+and/or ValidatingAdmissionWebhooks. These admission webhooks are registered
+through creating MutatingWebhookConfiguration or ValidatingWebhookConfiguration
+objects. Any policy enforcement is not in place until these webhook
+configurations are created, thereby registering the webhook. This creates a gap
+in enforcement spanning from when the kube-apiserver is started to until the
+webhook configuration objects are created and picked up by the dynamic admission
+controller. Another gap is the inability of the cluster administrator to protect
+against deletion of these webhook configuration objects as
+MutatingWebhookConfiguration and ValidatingWebhookConfiguration objects are not
+subject to webhook admission policy.
+
+This KEP aims to address these issues.  
+
+### Goals
+
+- A robust admission webhook registration process where there is no period of
+  time between the kube-apiserver coming up and the registration of an admission
+  webhook.
+
+- The registration of webhooks registered through this process should be
+  protected from alteration by API requests made by cluster users.
+
+- The new registration process should also have a highly privileged means of
+  being able to alter webhook registrations.
+ 
+## Proposal
+
+In short, this proposal is about augmenting `AdmissionConfig`'s plugin
+`configuration` [specification](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#authenticate-apiservers)
+to include a path to a configuration file containing one of the following:
+- `admissionregistration.k8s.io/v1.ValidatingWebhookConfigurationList` 
+- `admissionregistration.k8s.io/v1.MutatingWebhookConfigurationList` 
+- v1.List with items of admissionregistration.k8s.io/v1.ValidatingWebhookConfiguration 
+- v1.List  with items of admissionregistration.k8s.io/v1.MutatingWebhookConfiguration.
+This configuration is loaded  as static webhooks in api servers. These
+webhooks are called by validating and mutating admission plugins along with
+dynamically loaded webhooks for relevant admission requests. This also means
+that these webhooks are not API visible objects and hence cannot be modified by
+an API user.
+
+### New AdmissionConfig schema
+
+A `staticConfigFile` field is added to `configuration` field of a plugin object in
+`AdmissionConfig`.
+
+```yaml
+apiVersion: apiserver.config.k8s.io/v1
+kind: AdmissionConfiguration
+plugins:
+- name: ValidatingAdmissionWebhook
+  configuration:
+    ....
+    staticConfigFile: "<path-to-static-config-file>"
+- name: MutatingAdmissionWebhook
+  configuration:
+    ....
+    staticConfigFile: "<path-to-static-config-file>"
+```
+
+### Reconfiguring static webhook configuration
+
+The static configuration file is watched so that the static webhook
+configuration can be changed by editing the contents of the file.
```

Comment:
The idea of a dynamic, static webhook configuration seems somewhat unusual.  The forces that drew you to this solution (per the motivation) were related to avoiding a gap in enforcement and preventing users from mutating the values.  This prevents mutation, but doesn't solve the gap in enforcement.  If you need to change the file (hopefully infrequent), why wouldn't you simply restart the process at the same time.  The degree of access seems comparable.

Reply:
Watching the file for changes allows for updating other parts of the webhook configuration like eg. the matching criteria for performance reasons without causing unavailability of the API server. This is the motivation of allowing this.

Reply:
It's easy to imagine multiple categories of policy webhooks which would be placed in this file:

1. ones which are always present with a fixed config (would be present before the server is started, are guaranteed effective)
2. ones which have a variable config (typically related to intercepting as few resources as possible for performance reasons), but are not intended to be modifiable via the API

Handling updates to this file seems similar to reacting to CA bundle updates without requiring a server restart.